x-pack/solutions/security/plugins/security_solution/server/lib/detection_engine/rule_management/logic/detection_rules_client/utils.ts

```diff
@@ -61,3 +61,33 @@ export class RuleResponseValidationError extends Error {
     this.ruleId = ruleId;
   }
 }
+
+/**
+ * Given a rule from the file system and the set of installed rules this will merge the exception lists
+ * from the installed rules onto the rules from the file system.
+ * @param latestPrebuiltRule The latest prepackaged rule version that might have exceptions_lists
+ * @param existingRule The installed rules which might have user driven exceptions_lists
+ */
+export const mergeExceptionLists = (
+  latestPrebuiltRule: RuleResponse,
+  existingRule: RuleResponse
+): RuleResponse => {
+  if (latestPrebuiltRule.exceptions_list != null) {
+    if (existingRule.exceptions_list != null) {
+      const installedExceptionList = existingRule.exceptions_list;
+      const fileSystemExceptions = latestPrebuiltRule.exceptions_list.filter((potentialDuplicate) =>
+        installedExceptionList.every((item) => item.list_id !== potentialDuplicate.list_id)
+      );
+      return {
+        ...latestPrebuiltRule,
+        exceptions_list: [...fileSystemExceptions, ...existingRule.exceptions_list],
```

Comment:
spreading the rule object repeatedly is one reason why the performance of these routes isn't great. We should consider moving to a mutation based flow instead of copying the rule each time we modify it in memory. I don't think we should do that in this PR, because mixing the copy- and mutation-based paradigms will be more confusing, but we should fix it separately.

Reply:
Shifting to mutations is very risky. Side effects introduced at different levels can be extremely difficult to spot. For example, setting a reference to parts of a prebuilt rule within a mutable rule object can lead to accidental mutations of the prebuilt rule assets down the line. These kinds of bugs are notoriously hard to track and debug. Saving a few milliseconds of runtime at the cost of reliability and maintainability might not be a good trade-off. Paradigm shifts like this should be thoroughly discussed, and all risks clearly understood by the team before moving forward with implementation.